library/Cube/IcingaDb/IcingaDbCube.php

```diff
@@ -0,0 +1,209 @@
+     * and
+     *
+     * @return array
+     */
+    abstract public function getAvailableFactColumns();
+
+    /**
+     * @return Query
+     */
+    abstract public function prepareInnerQuery();
+
+    /**
+     * Get our inner query
+     *
+     * Hint: mostly used to get rid of NULL values
+     *
+     * @return Query
+     */
+    public function innerQuery()
+    {
+        if ($this->innerQuery === null) {
+            $this->innerQuery = $this->prepareInnerQuery();
+        }
+
+        return $this->innerQuery;
+    }
+
+    /**
+     * Get our rollup query
+     *
+     * @return Select
+     */
+    protected function rollupQuery()
+    {
+        if ($this->rollupQuery === null) {
+            $this->rollupQuery = $this->prepareRollupQuery();
+        }
+
+        return $this->rollupQuery;
+    }
+
+    /**
+     * Add a specific named dimension
+     *
+     * @param string $name
+     * @return $this
+     */
+    public function addDimensionByName($name)
+    {
+        $this->addDimension($this->createDimension($name));
+
+        return $this;
+    }
+
+    /**
+     * We first prepare the queries and to finalize it later on
+     *
+     * This way dimensions can be added one by one, they will be allowed to
+     * optionally join additional tables or apply other modifications late
+     * in the process
+     *
+     * @return void
+     */
+    protected function finalizeInnerQuery()
+    {
+        $innerQuery = $this->prepareInnerQuery();
+        $resolver = $innerQuery->getResolver();
+        $dimensions = $this->listDimensions();
+        $sourceTable = $innerQuery->getModel()->getTableName();
+
+        $groupBy = [];
+        foreach ($resolver->resolveRelations($sourceTable . '.vars') as $relation) {
+            foreach ($dimensions as $name => $dimension) {
+                $quotedDimension = $this->getDb()->quoteIdentifier($name);
+                $this->relation = $relation;
+                $dimension->addToCube($this);
+
+                $columns[$quotedDimension] = 'f.' . $quotedDimension;
+                $groupBy[] = $quotedDimension;
+            }
+        }
+
+        foreach ($resolver->resolveRelations($sourceTable . '.vars') as $relation) {
+            foreach ($this->listSlices() as $name) {
+                $this->relation = $relation;
+                $dimension = $this->createDimension($name);
+                $this->isSliced = true;
+                $dimension->addToCube($this);
+            }
+        }
+
+        if (! empty($groupBy)) {
+            $groupBy[count($groupBy) - 1] .= ' WITH ROLLUP';
+        }
+
+        $this->groupBy = $groupBy;
+
+        $availableFacts = $this->getAvailableFactColumns();
+        foreach ($this->chosenFacts as $alias) {
+            $columns[$alias] = 'SUM(f.' . $availableFacts[$alias] . ')';
+        }
+
+        $this->columns = $columns;
+    }
+
+    protected function prepareRollupQuery()
```

Comment:
Please re-think the naming here. The method calls it the rollup query, internally outer query is used. The same applies to `prepareFullQuery` where the full query is called rollup query internally. :thinking: